src/pip/_internal/req/req_install.py

```diff
@@ -553,7 +555,16 @@ def metadata(self) -> Any:
         return self._metadata
 
     def get_dist(self) -> BaseDistribution:
-        return get_directory_distribution(self.metadata_directory)
+        if self.metadata_directory:
+            return get_directory_distribution(self.metadata_directory)
+        elif self.local_file_path and self.is_wheel:
+            return get_wheel_distribution(
+                FilesystemWheel(self.local_file_path), canonicalize_name(self.name)
+            )
+        raise AssertionError(
+            f"InstallRequirement {self} has no metadata directory and no wheel: "
+            f"can't make a distribution."
+        )
```

Comment:
Does this show a user-facing succinct error message, or a traceback by default?

Reply:
It gives a traceback, which I think it is the right thing to do as it is a genuine bug if we reach this spot.